### ISSUE TITLE
Section#update broken when section contains anything other than body items

### ISSUE THREAD

**Issue by consp1racy** [lisawray/groupie#151]
This is the problem:

        // Dummy section to give us access to the flattened list of items in the new groups.
        final Section2 section = new Section2(groups);

The dummy section doesn't contain headers, footers, placeholder, empty state behavior so it behaves differently and indices don't line up.

I'd expect something along the lines of this:

    public void update(@NonNull final Collection<? extends Group> groups) {
        // Dummy section to give us access to the flattened list of items in the new groups.
        final Section2 section = new Section2(groups);

        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return getBodyItemCount();
            }

            @Override
            public int getNewListSize() {
                return section.getItemCount();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                Item oldItem = getItem(getHeaderItemCount() + oldItemPosition);
                Item newItem = section.getItem(newItemPosition);
                return newItem.isSameAs(oldItem);
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                Item oldItem = getItem(getHeaderItemCount() + oldItemPosition);
                Item newItem = section.getItem(newItemPosition);
                return newItem.equals(oldItem);
            }
        });

        super.removeAll(groups);
        children.clear();
        children.addAll(groups);
        super.addAll(groups);
        diffResult.dispatchUpdatesTo(listUpdateCallback);
        refreshEmptyState();
    }

    private ListUpdateCallback listUpdateCallback = new ListUpdateCallback() {
        @Override
        public void onInserted(int position, int count) {
            notifyItemRangeInserted(getHeaderItemCount() + position, count);
        }

        @Override
        public void onRemoved(int position, int count) {
            notifyItemRangeRemoved(getHeaderItemCount() + position, count);
        }

        @Override
        public void onMoved(int fromPosition, int toPosition) {
            notifyItemMoved(getHeaderItemCount() + fromPosition, getHeaderItemCount() + toPosition);
        }

        @Override
        public void onChanged(int position, int count, Object payload) {
            notifyItemRangeChanged(getHeaderItemCount() + position, count);
        }
    };

Basically offset calculations by the header item count, diff only body items, and then call `refreshEmptyState` to do the rest.

Do you think this is the correct way?

**Comment by consp1racy** [comment]
Ok, there are some errors above, most notably `getBodyItemCount()` returns placeholder items count when there are not content items. I finally settled on this version:

https://gist.github.com/consp1racy/0d70656167d9b46367c8deb4018c93c0

**Comment by lisawray** [comment]
Yes, you are right, this was a big oversight. I think I was blindsided because I ported the functionality from UpdatingGroup, which had no headers or footers. Let me see if I can get some failing tests going.
